File: internal/server/handlers/log_analytics.go
package handlers

import (
	"fmt"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/bestruirui/octopus/internal/op"
	"github.com/bestruirui/octopus/internal/server/middleware"
	"github.com/bestruirui/octopus/internal/server/resp"
	"github.com/bestruirui/octopus/internal/server/router"
	"github.com/gin-gonic/gin"
)

func init() {
	router.NewGroupRouter("/api/v1/log/analytics").
		Use(middleware.Auth()).
		AddRoute(router.NewRoute("/summary", http.MethodGet).Handle(getUsageAnalyticsSummary)).
		AddRoute(router.NewRoute("/timeseries", http.MethodGet).Handle(getUsageAnalyticsTimeseries)).
		AddRoute(router.NewRoute("/breakdown", http.MethodGet).Handle(getUsageAnalyticsBreakdown)).
		AddRoute(router.NewRoute("/export", http.MethodGet).Handle(getUsageAnalyticsExport)).
		AddRoute(router.NewRoute("/dimensions", http.MethodGet).Handle(getUsageAnalyticsDimensions))
}

func getUsageAnalyticsSummary(c *gin.Context) {
	filter, ok := parseUsageAnalyticsFilter(c)
	if !ok {
		return
	}
	result, err := op.UsageAnalyticsSummaryGet(c.Request.Context(), filter)
	if err != nil {
		resp.Error(c, http.StatusInternalServerError, err.Error())
		return
	}
	resp.Success(c, result)
}

func getUsageAnalyticsTimeseries(c *gin.Context) {
	filter, ok := parseUsageAnalyticsFilter(c)
	if !ok {
		return
	}
	result, err := op.UsageAnalyticsTimeseriesGet(c.Request.Context(), filter)
	if err != nil {
		resp.Error(c, http.StatusInternalServerError, err.Error())
		return
	}
	resp.Success(c, result)
}

func getUsageAnalyticsBreakdown(c *gin.Context) {
	filter, ok := parseUsageAnalyticsFilter(c)
	if !ok {
		return
	}
	options, ok := parseUsageBreakdownOptions(c)
	if !ok {
		return
	}
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	pageSize, _ := strconv.Atoi(c.DefaultQuery("page_size", "20"))
	result, err := op.UsageAnalyticsBreakdownSearchGet(
		c.Request.Context(),
		filter,
		options.dimension,
		options.search,
		page,
		pageSize,
		options.sort,
		options.descending,
	)
	if err != nil {
		resp.Error(c, http.StatusInternalServerError, err.Error())
		return
	}
	resp.Success(c, result)
}

type usageBreakdownOptions struct {
	dimension  string
	search     string
	sort       string
	descending bool
}

func parseUsageBreakdownOptions(c *gin.Context) (usageBreakdownOptions, bool) {
	dimension := strings.TrimSpace(c.DefaultQuery("dimension", "site"))
	if !validUsageDimension(dimension) {
		resp.Error(c, http.StatusBadRequest, "invalid dimension")
		return usageBreakdownOptions{}, false
	}
	sortBy := strings.TrimSpace(c.DefaultQuery("sort", "request_count"))
	if !validUsageSort(sortBy) {
		resp.Error(c, http.StatusBadRequest, "invalid sort")
		return usageBreakdownOptions{}, false
	}
	descending, err := parseBoolQuery(c, "descending", true)
	if err != nil {
		resp.Error(c, http.StatusBadRequest, err.Error())
		return usageBreakdownOptions{}, false
	}
	return usageBreakdownOptions{
		dimension:  dimension,
		search:     strings.TrimSpace(c.Query("search")),
		sort:       sortBy,
		descending: descending,
	}, true
}

func getUsageAnalyticsDimensions(c *gin.Context) {
	filter, ok := parseUsageAnalyticsFilter(c)
	if !ok {
		return
	}
	dimension := strings.TrimSpace(c.DefaultQuery("dimension", "site"))
	if !validUsageDimension(dimension) {
		resp.Error(c, http.StatusBadRequest, "invalid dimension")
		return
	}
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	pageSize, _ := strconv.Atoi(c.DefaultQuery("page_size", "50"))
	result, err := op.UsageAnalyticsDimensionsGet(
		c.Request.Context(),
		filter,
		dimension,
		c.Query("search"),
		page,
		pageSize,
	)
	if err != nil {
		resp.Error(c, http.StatusInternalServerError, err.Error())
		return
	}
	resp.Success(c, result)
}

func parseUsageAnalyticsFilter(c *gin.Context) (op.UsageAnalyticsFilter, bool) {
	startTime, err := parseOptionalInt64(c.Query("start_time"))
	if err != nil {
		resp.Error(c, http.StatusBadRequest, "invalid start_time")
		return op.UsageAnalyticsFilter{}, false
	}
	endTime, err := parseOptionalInt64(c.Query("end_time"))
	if err != nil {
		resp.Error(c, http.StatusBadRequest, "invalid end_time")
		return op.UsageAnalyticsFilter{}, false
	}
	timezone := strings.TrimSpace(c.DefaultQuery("timezone", "UTC"))
	if _, err := time.LoadLocation(timezone); err != nil {
		resp.Error(c, http.StatusBadRequest, "invalid timezone")
		return op.UsageAnalyticsFilter{}, false
	}
	scope := op.UsageMetricScope(strings.TrimSpace(c.DefaultQuery("metric_scope", string(op.UsageMetricScopeRequest))))
	if scope != op.UsageMetricScopeRequest && scope != op.UsageMetricScopeAttempt {
		resp.Error(c, http.StatusBadRequest, "invalid metric_scope")
		return op.UsageAnalyticsFilter{}, false
	}

	siteIDs, err := parseUsageIntList(c.Query("site_ids"))
	if err != nil {
		resp.Error(c, http.StatusBadRequest, err.Error())
		return op.UsageAnalyticsFilter{}, false
	}
	accountIDs, err := parseUsageIntList(c.Query("site_account_ids"))
	if err != nil {
		resp.Error(c, http.StatusBadRequest, err.Error())
		return op.UsageAnalyticsFilter{}, false
	}
	channelIDs, err := parseUsageIntList(c.Query("channel_ids"))
	if err != nil {
		resp.Error(c, http.StatusBadRequest, err.Error())
		return op.UsageAnalyticsFilter{}, false
	}
	apiKeyIDs, err := parseUsageIntList(c.Query("api_key_ids"))
	if err != nil {
		resp.Error(c, http.StatusBadRequest, err.Error())
		return op.UsageAnalyticsFilter{}, false
	}
	return op.UsageAnalyticsFilter{
		StartTime:       startTime,
		EndTime:         endTime,
		Timezone:        timezone,
		Scope:           scope,
		SiteIDs:         siteIDs,
		SiteAccountIDs:  accountIDs,
		ChannelIDs:      channelIDs,
		APIKeyIDs:       apiKeyIDs,
		RequestModels:   parseUsageStringList(c.Query("request_models")),
		ActualModels:    parseUsageStringList(c.Query("actual_models")),
		CanonicalModels: parseUsageStringList(c.Query("canonical_models")),
	}, true
}

func parseOptionalInt64(value string) (int64, error) {
	value = strings.TrimSpace(value)
	if value == "" {
		return 0, nil
	}
	return strconv.ParseInt(value, 10, 64)
}

func parseUsageIntList(value string) ([]int, error) {
	items := parseUsageStringList(value)
	result := make([]int, 0, len(items))
	for _, item := range items {
		id, err := strconv.Atoi(item)
		if err != nil || id <= 0 {
			return nil, fmt.Errorf("invalid id list")
		}
		result = append(result, id)
	}
	return result, nil
}

func parseUsageStringList(value string) []string {
	if strings.TrimSpace(value) == "" {
		return nil
	}
	seen := make(map[string]struct{})
	result := make([]string, 0)
	for _, item := range strings.Split(value, ",") {
		item = strings.TrimSpace(item)
		if item == "" {
			continue
		}
		if _, ok := seen[item]; ok {
			continue
		}
		seen[item] = struct{}{}
		result = append(result, item)
	}
	return result
}

func validUsageDimension(value string) bool {
	switch value {
	case "site", "site_account", "channel", "api_key", "request_model", "actual_model", "canonical_model":
		return true
	default:
		return false
	}
}

func validUsageSort(value string) bool {
	switch value {
	case "request_count", "attempt_count", "total_tokens", "cost", "success_rate", "duration":
		return true
	default:
		return false
	}
}

File: web/src/api/endpoints/log-analytics.ts
import { keepPreviousData, useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiClient, downloadApiFile } from '../client';

export type UsageMetricScope = 'request' | 'attempt';
export type UsageDimension =
    | 'site'
    | 'site_account'
    | 'channel'
    | 'api_key'
    | 'request_model'
    | 'actual_model'
    | 'canonical_model';

export interface UsageAnalyticsFilters {
    start_time?: number;
    end_time?: number;
    timezone: string;
    metric_scope: UsageMetricScope;
    site_ids?: number[];
    site_account_ids?: number[];
    channel_ids?: number[];
    api_key_ids?: number[];
    request_models?: string[];
    actual_models?: string[];
    canonical_models?: string[];
}

export interface UsageAnalyticsMetrics {
    metric_count: number;
    request_count: number;
    attempt_count: number;
    success_count: number;
    failed_count: number;
    canceled_count: number;
    indeterminate_count: number;
    success_rate: number;
    input_tokens: number;
    output_tokens: number;
    cache_read_tokens: number;
    cache_write_tokens: number;
    total_tokens: number;
    cost_usd: number;
    average_duration_ms: number;
    p95_duration_ms: number;
    average_ftut_ms: number;
    p95_ftut_ms: number;
    ftut_samples: number;
}

export interface UsageAnalyticsSummary extends UsageAnalyticsMetrics {
    scope: UsageMetricScope;
    start_time: number;
    end_time: number;
    timezone: string;
    drilldown_available: boolean;
    earliest_relay_log_time?: number | null;
}

export interface UsageTimeseriesPoint extends UsageAnalyticsMetrics {
    bucket_start: number;
    label: string;
}

export interface UsageTimeseries {
    scope: UsageMetricScope;
    granularity: 'hour' | 'day';
    timezone: string;
    points: UsageTimeseriesPoint[];
}

export interface UsageBreakdownItem extends UsageAnalyticsMetrics {
    id: number;
    name: string;
}

export interface UsageBreakdown {
    scope: UsageMetricScope;
    dimension: UsageDimension;
    page: number;
    page_size: number;
    total: number;
    items: UsageBreakdownItem[];
}

export type UsageBreakdownSort =
    | 'request_count'
    | 'attempt_count'
    | 'total_tokens'
    | 'cost'
    | 'success_rate'
    | 'duration';

export interface UsageBreakdownOptions {
    page: number;
    pageSize: 10 | 20 | 50;
    sort: UsageBreakdownSort;
    descending: boolean;
    search?: string;
}

export interface UsageAnalyticsExportOptions {
    filters: UsageAnalyticsFilters;
    dimension: UsageDimension;
    sort: UsageBreakdownSort;
    descending: boolean;
    search?: string;
}

export interface UsageDimensionOption {
    id: number;
    name: string;
}

export interface UsageDimensionsResult {
    dimension: UsageDimension;
    page: number;
    page_size: number;
    has_more: boolean;
    items: UsageDimensionOption[];
}

export interface RelayLogRepairPreview {
    rule_version: string;
    audit_id: string;
    matched: number;
    excluded: number;
    reasons: Record<string, number>;
    samples: Array<{
        id: number;
        time: number;
        model: string;
        channel: string;
        output_tokens: number;
    }>;
}

export interface RelayLogRepairResult extends RelayLogRepairPreview {
    batch_id: string;
    updated: number;
}

function appendArray(params: URLSearchParams, key: string, values?: Array<number | string>) {
    if (values?.length) params.set(key, values.join(','));
}

function analyticsSearch(filters: UsageAnalyticsFilters) {
    const params = new URLSearchParams();
    if (filters.start_time) params.set('start_time', String(filters.start_time));
    // UI date filters include the selected end second, while analytics
    // queries use a half-open [start, end) interval.
    if (filters.end_time) params.set('end_time', String(filters.end_time + 1));
    params.set('timezone', filters.timezone);
    params.set('metric_scope', filters.metric_scope);
    appendArray(params, 'site_ids', filters.site_ids);
    appendArray(params, 'site_account_ids', filters.site_account_ids);
    appendArray(params, 'channel_ids', filters.channel_ids);
    appendArray(params, 'api_key_ids', filters.api_key_ids);
    appendArray(params, 'request_models', filters.request_models);
    appendArray(params, 'actual_models', filters.actual_models);
    appendArray(params, 'canonical_models', filters.canonical_models);
    return params;
}

export function useUsageAnalyticsSummary(filters: UsageAnalyticsFilters) {
    return useQuery({
        queryKey: ['logs', 'analytics', 'summary', filters],
        queryFn: () =>
            apiClient.get<UsageAnalyticsSummary>(
                `/api/v1/log/analytics/summary?${analyticsSearch(filters).toString()}`,
            ),
        placeholderData: keepPreviousData,
        refetchInterval: 30000,
    });
}

export function useUsageAnalyticsTimeseries(filters: UsageAnalyticsFilters) {
    return useQuery({
        queryKey: ['logs', 'analytics', 'timeseries', filters],
        queryFn: () =>
            apiClient.get<UsageTimeseries>(
                `/api/v1/log/analytics/timeseries?${analyticsSearch(filters).toString()}`,
            ),
        placeholderData: keepPreviousData,
        refetchInterval: 30000,
    });
}

export function useUsageAnalyticsBreakdown(
    filters: UsageAnalyticsFilters,
    dimension: UsageDimension,
    options: UsageBreakdownOptions,
) {
    return useQuery({
        queryKey: ['logs', 'analytics', 'breakdown', filters, dimension, options],
        queryFn: () => {
            const params = analyticsSearch(filters);
            params.set('dimension', dimension);
            params.set('page', String(options.page));
            params.set('page_size', String(options.pageSize));
            params.set('sort', options.sort);
            params.set('descending', String(options.descending));
            if (options.search?.trim()) params.set('search', options.search.trim());
            return apiClient.get<UsageBreakdown>(
                `/api/v1/log/analytics/breakdown?${params.toString()}`,
            );
        },
        placeholderData: keepPreviousData,
        refetchInterval: 30000,
    });
}

export function useUsageAnalyticsDimensions(
    filters: UsageAnalyticsFilters,
    dimension: UsageDimension,
    search: string,
    page: number,
    pageSize = 20,
    enabled = true,
) {
    return useQuery({
        queryKey: ['logs', 'analytics', 'dimensions', filters, dimension, search, page, pageSize],
        queryFn: () => {
            const params = analyticsSearch(filters);
            params.set('dimension', dimension);
            params.set('search', search.trim());
            params.set('page', String(page));
            params.set('page_size', String(pageSize));
            return apiClient.get<UsageDimensionsResult>(
                `/api/v1/log/analytics/dimensions?${params.toString()}`,
            );
        },
        placeholderData: keepPreviousData,
        enabled,
    });
}

export function useUsageAnalyticsExport() {
    return useMutation({
        mutationFn: async (options: UsageAnalyticsExportOptions) => {
            const params = analyticsSearch(options.filters);
            params.set('dimension', options.dimension);
            params.set('sort', options.sort);
            params.set('descending', String(options.descending));
            if (options.search?.trim()) params.set('search', options.search.trim());
            return downloadApiFile(
                `/api/v1/log/analytics/export?${params.toString()}`,
                'octopus-usage.csv',
            );
        },
    });
}

export function useRelayLogRepairPreview() {
    return useMutation({
        mutationFn: (filter: { start_time?: number; end_time?: number }) =>
            apiClient.post<RelayLogRepairPreview>('/api/v1/log/repair/preview', filter),
    });
}

export function useRelayLogRepairExecute() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: (filter: { start_time?: number; end_time?: number }) =>
            apiClient.post<RelayLogRepairResult>('/api/v1/log/repair/execute', {
                ...filter,
                confirm: true,
            }),
        onSuccess: () => {
            // 不整体 invalidate ['logs']：infinite query 会被按序重取所有已加载页。
            queryClient.invalidateQueries({ queryKey: ['logs', 'head'] });
            queryClient.invalidateQueries({ queryKey: ['logs', 'analytics'] });
            queryClient.invalidateQueries({ queryKey: ['logs', 'site-action-targets'] });
            queryClient.removeQueries({ queryKey: ['logs', 'infinite'] });
        },
    });
}
